### PR TITLE
Allow setting scalar affine constraint functions via the modification API

### LINKED ISSUE #32: Supporting constraint function replacement
I noticed that it's not currently possible to replace the constraint function in an LQOI-based optimizer: 

```julia
optimizer = GLPK.GLPKOptimizerLP()
x = MOI.addvariables!(optimizer, 2)
f = MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.([1., 2.], x), 0.0)
c = MOI.addconstraint!(optimizer, f, MOI.LessThan(2.0))

f2 = MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.([2., 1.], x), 0.0)
MOI.set!(optimizer, MOI.ConstraintFunction(), c, f2)
```

gives: 

```
ArgumentError: ModelLike of type GLPK.GLPKOptimizerLP does not support setting the attribute `ConstraintFunction`

Stacktrace:
 [1] hidden_set!(::GLPK.GLPKOptimizerLP, ::MathOptInterface.ConstraintFunction, ::MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.LessThan{Float64}}, ::MathOptInterface.ScalarAffineFunction{Float64}) at /home/rdeits/locomotion/explorations/simple-qp-miqp/packages/v0.6/MathOptInterface/src/attributes.jl:588
 [2] set!(::GLPK.GLPKOptimizerLP, ::MathOptInterface.ConstraintFunction, ::MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.LessThan{Float64}}, ::MathOptInterface.ScalarAffineFunction{Float64}) at /home/rdeits/locomotion/explorations/simple-qp-miqp/packages/v0.6/MathOptInterface/src/attributes.jl:279
```

However, given that LQOI does support modifying individual coefficients of constraints, I think it ought to be possible to make this work at least when swapping out one affine function for another by writing a method that internally calls `modify!`. I haven't tried using `modify!` myself yet, but I'm willing to give it a try is my assumptions about this being possible and beneficial are correct.

### PR DESCRIPTION
**Pull request by rdeits**
Fixes #32 
Fixes https://github.com/JuliaOpt/GLPK.jl/issues/58

I'm not sure how to test this. There's a `solve_func_scalaraffine_lessthan` modification unit test in MOI.Test, but it doesn't seem to be run during the LQOI tests. I'd appreciate some help turning that on (and any other tests that are available).

### CODE REVIEW

**Comment by odow** [comment]
> I'd appreciate some help turning that on

The modification tests need to be added to LQOI. See GLPK for an example
https://github.com/JuliaOpt/GLPK.jl/blob/9a9e9f4ec6ac5bcc1b9b1b1e1a79fe9d2b43efe5/test/MOIWrapper.jl#L23-L25
The problem is that the mock solver in LQOI doesn't test the solution (since `solve=false`)
https://github.com/JuliaOpt/LinQuadOptInterface.jl/blob/d448b17367bd49f451255cae9934bb0550843673/test/runtests.jl#L10
so it isn't well tested. This needs to improve. I usually test by running GLPK and Gurobi on the changes.

@joaquimg is probably the person who understands the current state of the tests the best.

**Comment by rdeits** [comment]
Got it, thanks. I'll run glpk and gurobi with the modification tests on and open PRs to turn them on for those packages.

**Comment by rdeits** [comment]
By the way, https://github.com/JuliaOpt/Gurobi.jl/pull/125 restricts LQOI to version [0.1, 0.2), but LQOI master is version 0.2+. Is that restriction necessary?

**Comment by rdeits** [comment]
PRs opened: https://github.com/JuliaOpt/GLPK.jl/pull/56 https://github.com/JuliaOpt/Gurobi.jl/pull/134

**Comment by joaquimg** [comment]
The problem with testing solutions is that one need to gather all the intermediary solutions (objective, variable primal and dual and constraint primal and dual) from some solver like Gurobi.
The mock solver is "ready"(but not tested) to receive these vectors as input.
Doing that for the "linear tests" is strainghtforward, not sure about the unit tests, but should not be hard.

**Comment by rdeits** [comment]
Hm, i just realized this isn't quite right, as I'm not handling a possible modification of the constant in the affine function. It doesn't seem to be possible to use a `ScalarConstantChange` in LQOI. Should I just replace the constraint set at the same time?

**Comment by joaquimg** [comment]
You are planning to modify the sparsity pattern of the constraint?
About performance... I would guess its faster to delete the constraint and add a new than query the old constraint and modify all coefficients.
`ScalarConstantChange` should be use when modifying few coefficients IMO.

> Should I just replace the constraint set at the same time?

I would just replace the set.

**Comment by tkoolen** [comment]
Regarding testing modifications: I've been using the following in OSQP:

https://github.com/oxfordcontrol/OSQP.jl/blob/200811d9ba2ecafb07859764a271bcd3801c1b63/test/MathOptInterfaceOSQP.jl#L133-L176

Example usage:

https://github.com/oxfordcontrol/OSQP.jl/blob/200811d9ba2ecafb07859764a271bcd3801c1b63/test/MathOptInterfaceOSQP.jl#L249-L254

File: src/constraints/scalaraffine.jl [hunk 1]
```diff
@@ -169,6 +169,23 @@ function MOI.modify!(m::LinQuadOptimizer, c::LCI{S}, chg::MOI.ScalarCoefficientC
...
+    previous = MOI.get(m, attr, CI)
+    for term in previous.terms
+        var = term.variable_index
+        chg = MOI.ScalarCoefficientChange{Float64}(var, zero(Float64))
+        MOI.modify!(m, CI, chg)
+    end
+    for term in replacement.terms
+        var = term.variable_index
+        chg = MOI.ScalarCoefficientChange{Float64}(var, term.coefficient)
+        MOI.modify!(m, CI, chg)
```
Comment (tkoolen) [inline src/constraints/scalaraffine.jl]: This also doesn't seem to support duplicate coefficients.

File: src/constraints/scalaraffine.jl [hunk 2]
```diff
@@ -169,6 +169,24 @@ function MOI.modify!(m::LinQuadOptimizer, c::LCI{S}, chg::MOI.ScalarCoefficientC
...
+    end
+    change_rhs_coefficient!(m, m[CI], -replacement.constant)
```
Comment (tkoolen) [inline src/constraints/scalaraffine.jl]: Does this handle the constant encoded in the set properly (e.g. `MOI.EqualTo(1.0)`)?

**Comment by tkoolen** [comment]
Some thoughts:

* there should probably be an `LQOI.change_matrix_coefficients!` (plural) to reduce the number of low level API calls
* at least for Gurobi, it seems that [sparsity pattern changes are already handled internally](http://www.gurobi.com/documentation/8.0/refman/c_grbchgcoeffs.html). It's possible that simply zeroing everything (as opposed to just the currently-nonzero coefficients) is already efficient, so that `get`ting the constraint function from the solver is not necessary.
* something like OSQP's [`MatrixModificationCache`](https://github.com/oxfordcontrol/OSQP.jl/blob/200811d9ba2ecafb07859764a271bcd3801c1b63/src/modcaches.jl#L33) is probably needed to handle duplicate coefficients in an (at least algorithmically) efficient way (note that this is actually a moderately slow part of the OSQP interface due to lots of `Dict` lookups).

**Comment by odow** [comment]
What's the status of this? Is it necessary for the next release of LQOI? Or can it wait?

**Comment by rdeits** [comment]
Sorry, I haven't had time to vet this properly, and I think its handling of scalar affine functions is incorrect. It's on my todo list but not quite at the top. 

If you're OK with leaving this open, I'll let you know when it's ready for review.

**Comment by rdeits** [comment]
But to answer your question, no, this should not be necessary for the next release. It's just adding a new feature.

**Comment by rdeits** [comment]
@odow will this change also be rendered moot by your switch to using bridges here? I'm just trying to decide whether to keep this as a local modification for my own work or to actually put in the time to finish it correctly.

**Comment by odow** [comment]
I would keep it as a local change for now. Other than you, no one is clamoring for this change, so it doesn't make sense to spend a lot of time on it.

**Comment by rdeits** [comment]
Hey @odow this is actually becoming a bit more pressing for my work. I believe I've fixed the bug with RHS coefficients that was previously in this branch, and the lack of this feature is currently blocking https://github.com/tkoolen/Parametron.jl/pull/75

The scalar part of this PR is tested by https://github.com/JuliaOpt/GLPK.jl/pull/56 and the vector part is tested by https://github.com/tkoolen/Parametron.jl/pull/75

Would you be willing to consider merging this, even if it will eventually be replaced by the bridges?

**Comment by rdeits** [comment]
Sigh, nope, still more bugs with constraint constant handling on this branch. At least one of them will be fixed by #53

**Comment by odow** [comment]
> Would you be willing to consider merging this, even if it will eventually be replaced by the bridges?

Yes. Let me know when you want to.

**Comment by rdeits** [comment]
Ok, sounds good. I'll make sure all the bugs in master are fixed first, then add some more tests for this and let you know.

**Comment by odow** [comment]
I've merged #53. Will take a look at this

File: src/constraints/scalaraffine.jl [hunk 3]
```diff
@@ -196,18 +196,65 @@ function MOI.modify!(model::LinQuadOptimizer, index::LCI{S}, change::MOI.ScalarC
     change_matrix_coefficient!(model, row, column, change.new_coefficient)
 end
 
+function _replace_with_matching_sparsity!(model::LinQuadOptimizer, previous::Linear, replacement::Linear, row)
```
Comment (odow) [inline src/constraints/scalaraffine.jl]: I know the rest of the code doesn't, but can we add docstrings for these new functions to remind ourselves of why they are needed and what they do?

File: src/constraints/scalaraffine.jl [hunk 4]
```diff
@@ -196,18 +196,65 @@ function MOI.modify!(model::LinQuadOptimizer, index::LCI{S}, change::MOI.ScalarC
...
+    rows = fill(row, length(previous.terms))
+    cols = [model.variable_mapping[t.variable_index] for t in previous.terms]
+    coefs = fill(0.0, length(previous.terms))
+    change_matrix_coefficients!(model, rows, cols, coefs)
+
+    # Next, set the new constraint function terms
+    rows = fill(row, length(replacement.terms))
+    cols = [model.variable_mapping[t.variable_index] for t in replacement.terms]
+    coefs = MOIU.coefficient.(replacement.terms)
+    change_matrix_coefficients!(model, rows, cols, coefs)
```
Comment (odow) [inline src/constraints/scalaraffine.jl]: Do these account for duplicate coefficients? (Edit: looks from below that the inputs are canonicalized. All the more reason for docstrings.)

File: src/constraints/scalaraffine.jl [hunk 5]
```diff
@@ -196,18 +196,65 @@ function MOI.modify!(model::LinQuadOptimizer, index::LCI{S}, change::MOI.ScalarC
...
+    # and then set the replacement terms.
+    row = model[CI]
+    if _matching_sparsity_pattern(previous, replacement)
+        _replace_with_matching_sparsity!(model, previous, replacement, row)
+    else
+        _replace_with_different_sparsity!(model, previous, replacement, row)
+    end
+    change_rhs_coefficient!(model, row, get_rhs(model, row) - (replacement.constant - previous.constant))
+    model.constraint_constant[model[CI]] = replacement.constant
+    nothing
```
Comment (odow) [inline src/constraints/scalaraffine.jl]: Nit: `return` or `return nothing`

File: src/constraints/vectoraffine.jl [hunk 6]
```diff
@@ -76,6 +76,62 @@ function MOI.modify!(model::LinQuadOptimizer, index::VLCI{<: VecLinSets},
     end
 end
 
+_constant(f::Linear) = f.constant
+_constant(f::VecLin) = f.constants
```
Comment (odow) [inline src/constraints/vectoraffine.jl]: Are there MOIU functions for this?

File: src/constraints/scalaraffine.jl (hunk 3 above)
Comment (rdeits) [reply src/constraints/scalaraffine.jl]: done

File: src/constraints/vectoraffine.jl (hunk 6 above)
Comment (rdeits) [reply src/constraints/vectoraffine.jl]: There are, but they're likewise `_`-prefixed so I didn't want to use them externally. `MOIU` does have `constant()` but it unnecessarily allocates a vector for scalar functions.

**Comment by joaquimg** [comment]
Looks good to go. I just added one question about the sign change.

File: src/constraints/scalaraffine.jl [hunk 7]
```diff
@@ -183,7 +183,7 @@ function MOI.get(model::LinQuadOptimizer, ::MOI.ConstraintFunction, index::LCI{<
         model.variable_references[columns],
         coefficients
     )
-    Linear(terms, -model.constraint_constant[row])
+    Linear(terms, model.constraint_constant[row])
```
[the remark this answers is not in the document]
Comment (rdeits) [reply src/constraints/scalaraffine.jl]: Nope. But I actually already made that fix (and added a test) in https://github.com/JuliaOpt/LinQuadOptInterface.jl/pull/53/files#diff-c1cce28d94a3c050507f8711d20ca7b2R186

**Comment by tkoolen** [comment]
Ready to go?

**Comment by rdeits** [comment]
Yeah, I think this should be ready. Can we merge it?

File: src/constraints/vectoraffine.jl [hunk 8]
```diff
@@ -76,6 +76,101 @@ function MOI.modify!(model::LinQuadOptimizer, index::VLCI{<: VecLinSets},
...
+    _matching_sparsity_pattern(f1::T, f2::T) where {T <: Union{Linear, VecLin}}
+
+Internal function, not intended for external use.
+
+Determines whether functions `f1` and `f2` have the same sparsity pattern
+w.r.t. their constraint rows and columns. Assumes both functions are already
+in canonical form.
+"""
+function _matching_sparsity_pattern(f1::T, f2::T) where {T <: Union{Linear, VecLin}}
+    indices(f1.terms) == indices(f2.terms) || return false
```
Comment (tkoolen) [inline src/constraints/vectoraffine.jl]: `indices` -> `Compat.axes`.

File: src/solver_interface.jl [hunk 9]
```diff
@@ -204,6 +204,24 @@ Set the linear coefficient of the variable in column `col`, constraint `row` to
...
+Change multiple linear coefficients simultaneously. Sets the linear coefficient
+of the variable at column `cols[i]` in constraint `rows[i]` to `coefs[i]` for
+`i in 1:length(rows)`. Requires that `length(rows) == length(cols) == length(coefs)`.
+
+By default, this method just calls
+`change_matrix_coefficient!(m, rows[i], cols[i], coefs[i])` repeatedly, but
+some solver interfaces may offer more efficient implementations.
+"""
+function change_matrix_coefficients!(m, rows, cols, coefs)
+    @boundscheck((indices(rows) == indices(cols) == indices(coefs)) || throw(DimensionMismatch()))
```
Comment (tkoolen) [inline src/solver_interface.jl]: `indices` -> `Compat.axes`.

File: src/constraints/vectoraffine.jl (hunk 8 above)
Comment (rdeits) [reply src/constraints/vectoraffine.jl]: done

File: src/solver_interface.jl (hunk 9 above)
Comment (rdeits) [reply src/solver_interface.jl]: done

**Comment by odow** [comment]
Merging this. We can rebase #60 on top of this and address any other compat issues.

**Pull request merged**